File: meta-android-app/app/src/test/java/com/meta/wearable/dat/externalsampleapps/cameraaccess/saispike/conversation/ConversationHarness.kt
/* sai-fi — voice concierge. */

// The closed loop, with everything real except the brain and the agent.
//
//     brain ──tool calls──▶ LiveTurnGate ──effects──▶ Concierge ──▶ HttpAgentBridge
//       ▲                        │                       │              │
//       │                   defer/inject                 │       ScriptedAgent (VoiceTransport)
//       │                        ▲                       ▼              │
//       └───nudges / [system]────┴── AgentEventRouter ◀── agent events ──┘
//
// Layer 1 (the golden catalog) fakes the brain; layer 2 (the eval) fakes the FSM and the agent. The
// failures that live in the JOIN are invisible to both, and they are the ones the on-device check
// calls its highest risk — "if Sai says 'I'll do that next' and then it never runs, the drain never
// fired". Nothing in either layer can catch that, because in one the drain is scripted and in the
// other there is no queue at all.
//
// What is REAL here: Concierge (the FSM), HttpAgentBridge and its transport seam, LiveTurnGate,
// LiveVoiceChannel, AgentEventRouter, HeldNudgeQueue, ActivityLog, and the effect-parse boundary
// (`applyClientEffects`) the model's tool calls actually go through.
//
// What is not: the model (a [Brain]) and the agent (a [ScriptedAgent]).
//
// The client-policy block below — activity log, nudge routing, mute holding — is deliberately a
// mirror of CallService's `onAgentEvent`. That code is welded to a Service and cannot be reached
// from a JVM test; keeping the mirror small and pointing at the original is the cheaper half of the
// trade. If the two drift, the on-device check is what notices, so any change there belongs here too.

package com.meta.wearable.dat.externalsampleapps.cameraaccess.saispike.conversation

import com.meta.wearable.dat.externalsampleapps.cameraaccess.saispike.ActivityLog
import com.meta.wearable.dat.externalsampleapps.cameraaccess.saispike.AgentEventRouter
import com.meta.wearable.dat.externalsampleapps.cameraaccess.saispike.GateAction
import com.meta.wearable.dat.externalsampleapps.cameraaccess.saispike.HeldNudgeQueue
import com.meta.wearable.dat.externalsampleapps.cameraaccess.saispike.HttpAgentBridge
import com.meta.wearable.dat.externalsampleapps.cameraaccess.saispike.LiveTurnGate
import com.meta.wearable.dat.externalsampleapps.cameraaccess.saispike.LiveVoiceChannel
import com.meta.wearable.dat.externalsampleapps.cameraaccess.saispike.NudgeAction
import com.meta.wearable.dat.externalsampleapps.cameraaccess.saispike.VoiceTransport
import com.meta.wearable.dat.externalsampleapps.cameraaccess.saispike.TaskRouting
import com.meta.wearable.dat.externalsampleapps.cameraaccess.saispike.agentEventToJson
import com.meta.wearable.dat.externalsampleapps.cameraaccess.saispike.userQuietMs
import com.meta.wearable.dat.externalsampleapps.cameraaccess.saispike.describePhoneClock
import com.meta.wearable.dat.externalsampleapps.cameraaccess.saispike.ClientBrain
import com.meta.wearable.dat.externalsampleapps.cameraaccess.saispike.fsm.AgentEvent
import com.meta.wearable.dat.externalsampleapps.cameraaccess.saispike.fsm.Concierge
import com.meta.wearable.dat.externalsampleapps.cameraaccess.saispike.fsm.ConciergeState
import org.json.JSONArray
import org.json.JSONObject

/** A line in the conversation, as a person listening to the call would experience it. */
data class Line(val speaker: String, val text: String)

/** Defaults mirroring CallService's. */
private const val DEFAULT_ASK_FIRST_MS = 45_000L
private const val STEP_FAILURE_NUDGE_INTERVAL_MS = 60_000L

class ConversationHarness(
    private val brain: Brain,
    val clock: HarnessClock = HarnessClock(),
    /** How long the model's speech occupies the channel. The window in which a nudge gets deferred. */
    var speakingMs: Long = 800L,
) {
  val log = mutableListOf<String>()

  /**
   * Mirrors this conversation to the presenter when `SAI_PRESENTER=1`, so a test can be watched in
   * the same browser view a real call uses. Null — and free — otherwise.
   */
  private val presenter = PresenterPublisher.fromEnvOrNull()

  private fun log(line: String) {
    log += line
    presenter?.log(line)
  }

  val gate = LiveTurnGate { clock.now }
  val activityLog = ActivityLog { clock.now }
  private val heldNudges = HeldNudgeQueue()

  val agent = ScriptedAgent(clock, deliver = { onAgentEvent(it) }, log = ::log)

  /** Everything the user would have heard, plus what the agent said, in order. */
  val transcript = mutableListOf<Line>()

  /** Every session-state projection the FSM published. */
  val sessionStates = mutableListOf<AgentEvent.SessionState>()

  var muted: Boolean = false
    private set

  private var lastUserSpeechAt = 0L
  /** Mirrors CallService's field of the same name — the quiet clock's stop point. See [userQuietMs]. */
  private var workStartedAt = 0L
  private var lastStepFailureNudgeAt = 0L

  /**
   * Note that work the user is waiting on has begun, keeping the FIRST such moment since they spoke.
   *
   * Mirrors `CallService.markWorkStarted`, minus the capture site it has no camera for.
   */
  private fun markWorkStarted() {
    if (workStartedAt < lastUserSpeechAt || workStartedAt == 0L) workStartedAt = clock.now
  }

  /**
   * Swapped for a [LiveAgent] when the contract tier runs.
   *
   * Indirected through a delegate rather than taken as a constructor argument because the scripted
   * agent needs the harness (to deliver events) and the harness needs the transport (to build the
   * bridge) — a cycle that a settable field breaks without making every scenario declare a transport
   * it does not care about.
   */
  private var transport: VoiceTransport? = null

  /** Send this run's traffic to [replacement] instead of the scripted agent. Call before [start]. */
  fun useTransport(replacement: VoiceTransport) {
    transport = replacement
  }

  private val bridge =
      HttpAgentBridge(
          "M-harness",
          object : VoiceTransport {
            override suspend fun sendMessage(
                machineId: String,
                message: String,
                attachments: JSONArray?,
                follow: Boolean,
            ) = (transport ?: agent).sendMessage(machineId, message, attachments, follow)

            override fun abandonTurn() = (transport ?: agent).abandonTurn()

            override suspend fun post(path: String, body: JSONObject): JSONObject =
                (transport ?: agent).post(path, body)
          },
          ::log,
      )

  /**
   * Hand an agent event to the client, as the turn-stream reader does.
   *
   * Public so a live transport, whose events arrive on its own reader rather than on the clock, can
   * put them through the same path the scripted one uses — the FSM, then the activity log, then the
   * nudge routing.
   */
  suspend fun deliverAgentEvent(event: AgentEvent) = onAgentEvent(event)

  /**
   * Leave a photo on the bridge for whatever writes next — what a real capture does when it lands.
   *
   * The stash belongs to the next write, which is exactly the immediate path's rule: the adapter
   * drains it on `forwardTask`. A held task takes its photos with it instead (see
   * `takePendingAttachments`), which is why a queued request cannot end up carrying someone else's.
   */
  fun stashAttachment(attachment: com.meta.wearable.dat.externalsampleapps.cameraaccess.saispike.fsm.TaskAttachment) =
      bridge.addPendingAttachment(attachment)

  /** Let the presenter show a beat of silence, when one is what a person would hear. */
  fun beat(ms: Long) = presenter?.pause(ms)

  val concierge =
      Concierge(
          agent = bridge,
          voice = LiveVoiceChannel { kind, text -> pushToModel(kind, text) },
          engine = ClientBrain,
          timer = clock,
          onSessionState = { s ->
            sessionStates += s
            activityLog.record(agentEventToJson(s))
          },
          now = { clock.now },
          log = ::log,
      )

  val state: ConciergeState
    get() = concierge.getState()

  /** What `getSaiStatus` would answer right now — the same renderer the device uses. */
  fun status(): String = activityLog.statusText()

  // ── Driving the conversation ─────────────────────────────────────────────────────────────────────

  /** Bring the Live session up. */
  suspend fun start() {
    presenter?.hello("harness")
    presenter?.state(active = true, status = "in a call", machine = "harness", muted = false)
    runGate(gate.onConnect())
    runGate(gate.onSetupComplete())
  }

  /** The user says something. */
  suspend fun user(utterance: String) {
    lastUserSpeechAt = clock.now
    // Speaking while we are already busy restarts the wait rather than a silence — they are plainly
    // here, and plainly still waiting. Mirrors CallService's rule in `transcript`.
    if (state.inFlight.isNotEmpty() || state.queue.isNotEmpty()) markWorkStarted()
    transcript += Line("you", utterance)
    presenter?.speak("you", utterance)
    gate.onUserTranscript(utterance)
    modelTurn(utterance)
  }

  /**
   * The user talks over the model.
   *
   * Two things happen on a real barge-in and both matter: the server VAD raises `interrupted`, which
   * ends the model's turn and opens the discard window, and then the new utterance is a fresh turn.
   */
  suspend fun bargeIn(utterance: String) {
    runGate(gate.onInterrupted())
    presenter?.interrupted()
    log("— barge-in —")
    user(utterance)
  }

  /** Mute / unmute, which on the device is the temple button or the on-screen control. */
  suspend fun setMuted(value: Boolean) {
    muted = value
    presenter?.state(active = true, status = if (value) "muted" else "in a call", machine = "harness", muted = value)
    if (value) {
      runGate(gate.injectSessionState("muted", "[system] you are muted", sticky = true))
    } else {
      runGate(gate.injectSessionState("unmuted", "[system] you are unmuted", sticky = false))
      // Anything that finished while muted is offered now — once, not replayed as a pile.
      heldNudges.drain().forEach { runGate(gate.injectNudge(it.kind, it.nudge)) }
    }
  }

  /** A token-expiry reconnect mid-call. */
  suspend fun reconnect() {
    runGate(gate.onConnect())
    runGate(gate.onSetupComplete())
  }

  /** Let time pass, delivering whatever the agent had scheduled. */
  suspend fun advance(ms: Long) = clock.advance(ms)

  /** Run until the agent has nothing left to say. */
  suspend fun settle() = clock.drain()

  // ── The model's half ─────────────────────────────────────────────────────────────────────────────

  /**
   * Give the model a turn, and put whatever it decided through the real boundaries.
   *
   * The order matters and mirrors the device: the model's speech opens the turn (so a nudge arriving
   * now is deferred by the gate), its tool calls go through the same routing GeminiLiveClient
   * applies, and the turn closes [speakingMs] later — which is when anything held gets released.
   */
  private suspend fun modelTurn(input: String) {
    val turn = brain.turn(input, state)
    if (!turn.speech.isNullOrBlank()) {
      runGate(gate.onSaiTranscript(turn.speech))
      transcript += Line("sai", turn.speech)
      presenter?.speak("sai", turn.speech)
    }
    routeCalls(turn.calls)
    // The turn ends after the model has finished speaking, not instantly: the gap is the window in
    // which a completion landing mid-sentence is held, which is the race worth testing.
    //
    // WHICH clock matters, and getting it wrong is silent. A live transport delivers on real time and
    // nothing advances the virtual one, so scheduling the turn end there meant it never fired:
    // `modelSpeaking` stayed true for the rest of the call and every nudge was deferred behind a turn
    // that would never end. The model looked mute when it had simply never been told anything — the
    // same shape as the bug LiveTurnGate was extracted for, reproduced in the harness, and it spoiled
    // a live demo before the cause was spotted.
    val speakFor = if (turn.speech.isNullOrBlank()) 0L else speakingMs
    if (transport != null) {
      kotlinx.coroutines.delay(speakFor)
      runGate(gate.onGenerationOrTurnEnd(generationEnded = false, turnEnded = true))
    } else {
      clock.scheduleSuspending(speakFor) {
        runGate(gate.onGenerationOrTurnEnd(generationEnded = false, turnEnded = true))
      }
    }
  }

  /** The tool-call half of GeminiLiveClient.handleToolCall, minus the socket. */
  private suspend fun routeCalls(calls: JSONArray) {
    // Mirrors the real client: the gate is told a turn is in flight BEFORE any effect reaches the
    // FSM, because the FSM answers some of them by speaking and a spoken line is a client turn.
    gate.onToolCall()
    val effects = JSONArray()
    val hasCapture =
        (0 until calls.length()).any { calls.getJSONObject(it).optString("name") == "captureImage" }
    for (i in 0 until calls.length()) {
      val c = calls.getJSONObject(i)
      val name = c.optString("name")
      val effect = fcToEffect(c)
      when (name) {
        "getSaiStatus" -> log("→ tool: getSaiStatus → ${status()}")
        "getLocalTime" -> log("→ tool: getLocalTime → ${describePhoneClock()}")
        "forwardToAgent",
        "enqueue",
        "relayToAgent" -> {
          val wantsPhoto = c.optJSONObject("args")?.optBoolean("attachLatestImage") == true
          when (val routing = gate.routeTaskCall(name, effect, wantsPhoto, hasCapture)) {
            is TaskRouting.HeldForPhoto -> log(routing.log)
            is TaskRouting.Emit -> {
              effects.put(effect)
              log(routing.log)
            }
          }
        }
        else -> {
          effects.put(effect)
          log("→ effect: $name")
        }
      }
    }
    if (effects.length() > 0) {
      markWorkStarted()
      concierge.applyClientEffects(effects)
    }
  }

  /** `{ kind: name, ...args }` — the concierge effect shape, as GeminiLiveClient builds it. */
  private fun fcToEffect(fc: JSONObject): JSONObject {
    val eff = JSONObject().put("kind", fc.optString("name"))
    fc.optJSONObject("args")?.let { args -> args.keys().forEach { eff.put(it, args.get(it)) } }
    return eff
  }

  /** The FSM's voice out, wired the way CallService wires it: everything goes through the gate. */
  private fun pushToModel(kind: String, text: String) {
    // Queued rather than run inline: `say` is called from inside the FSM's mutex, and giving the
    // model a turn here would re-enter it.
    clock.scheduleSuspending(0) { runGate(gate.injectNudge(kind, text)) }
  }

  // ── The client's own policy on agent events (mirrors CallService.onAgentEvent) ────────────────────

  private suspend fun onAgentEvent(e: AgentEvent) {
    // Mirrors the gate in VoiceSession's reader, and covers all three sinks for the same reason it
    // does: the tail of an aborted turn is not spoken, not written to the log `getSaiStatus` answers
    // from, and not given to the FSM. Two of those three are nothing to do with the FSM, which is why
    // suppressing its reaction alone left the result still being read out.
    if (concierge.disownsAgentEvents()) {
      log("dropped ${e::class.simpleName} from an aborted turn")
      return
    }
    concierge.handleAgentEvent(e)

    val json = agentEventToJson(e)
    activityLog.record(json)
    if (e is AgentEvent.Complete || e is AgentEvent.Error) {
      transcript += Line("agent", json.optString("summary").ifEmpty { json.optString("text") })
    }

    val action =
        AgentEventRouter.route(
            event = json,
            muted = muted,
            // The shipped function, not a copy of it. This was an inlined `when` that had already
            // drifted from the one CallService uses, which is the whole hazard a mirror carries.
            userQuietMs = userQuietMs(clock.now, lastUserSpeechAt, workStartedAt),
            askFirstThresholdMs = DEFAULT_ASK_FIRST_MS,
            sinceLastStepFailureMs =
                if (lastStepFailureNudgeAt == 0L) Long.MAX_VALUE
                else clock.now - lastStepFailureNudgeAt,
            stepFailureIntervalMs = STEP_FAILURE_NUDGE_INTERVAL_MS,
        )
    when (action) {
      is NudgeAction.Ignore -> {}
      is NudgeAction.Drop -> log("→ nudge: ${action.why}")
      is NudgeAction.InjectStepFailure -> {
        lastStepFailureNudgeAt = clock.now
        runGate(gate.injectNudge("step-failed", action.nudge))
      }
      is NudgeAction.Inject -> runGate(gate.injectNudge(action.kind, action.nudge))
      is NudgeAction.Hold ->
          if (!heldNudges.add(action.kind, action.nudge)) {
            log("→ nudge: ${action.kind} — discarded while muted (stale by the time it could be heard)")
          }
    }
  }

  // ── The gate's action interpreter ────────────────────────────────────────────────────────────────

  /**
   * Perform what the gate decided — the same job GeminiLiveClient.run does, except that a turn sent
   * to the model comes back here as another model turn, which is what closes the loop.
   */
  private suspend fun runGate(actions: List<GateAction>) {
    for (action in actions) {
      when (action) {
        is GateAction.Log -> log(action.text)
        is GateAction.SendTurn -> modelTurn(action.text)
        is GateAction.SaiTranscript -> {}
        is GateAction.UserTranscript -> {}
        is GateAction.TurnComplete -> {}
        is GateAction.FlushPlayback -> {}
      }
    }
  }

  // ── Assertion helpers ────────────────────────────────────────────────────────────────────────────

  /** Everything the user heard, lowercased and joined — for "was this ever said" checks. */
  fun heard(): String = transcript.filter { it.speaker == "sai" }.joinToString(" | ") { it.text }.lowercase()

  fun saidSomethingLike(vararg fragments: String) = fragments.any { heard().contains(it.lowercase()) }

  fun logHas(fragment: String) = log.any { it.contains(fragment) }
}

File: meta-ios-app (untested on-device)/SaiFiCore/Sources/SaiFiCore/Conversation/ConversationHarness.swift
/* sai-fi — voice concierge. */

// The closed loop, with everything real except the brain and the agent.
//
//     brain ──tool calls──▶ LiveTurnGate ──effects──▶ Concierge ──▶ HttpAgentBridge
//       ▲                        │                       │              │
//       │                   defer/inject                 │       ScriptedAgent (VoiceTransport)
//       │                        ▲                       ▼              │
//       └───nudges / [system]────┴── AgentEventRouter ◀── agent events ──┘
//
// Layer 1 (the golden catalog) fakes the brain; layer 2 (the eval) fakes the FSM and the agent. The
// failures that live in the JOIN are invisible to both, and they are the ones the on-device check
// calls its highest risk — "if Sai says 'I'll do that next' and then it never runs, the drain never
// fired". Nothing in either layer can catch that, because in one the drain is scripted and in the
// other there is no queue at all.
//
// What is REAL here: Concierge (the FSM), HttpAgentBridge and its transport seam, LiveTurnGate,
// LiveVoiceChannel, AgentEventRouter, HeldNudgeQueue, ActivityLog, and the effect-parse boundary
// (`applyClientEffects`) the model's tool calls actually go through.
//
// What is not: the model (a Brain) and the agent (a ScriptedAgent).
//
// The client-policy block below — activity log, nudge routing, mute holding — is deliberately a
// mirror of CallService's `onAgentEvent`. That code is welded to a Service and cannot be reached
// from a JVM test; keeping the mirror small and pointing at the original is the cheaper half of the
// trade. If the two drift, the on-device check is what notices, so any change there belongs here too.
//
// Ported from Android `conversation/ConversationHarness.kt`. The presenter feed is out of scope.

import Foundation
import os

/// A line in the conversation, as a person listening to the call would experience it.
public struct Line: Sendable, Equatable {
  public let speaker: String
  public let text: String
}

/// Defaults mirroring CallService's.
private let defaultAskFirstMs: Int64 = 45_000
private let stepFailureNudgeIntervalMs: Int64 = 60_000

private final class TransportBox: VoiceTransport, @unchecked Sendable {
  var inner: (any VoiceTransport)?
  var fallback: (any VoiceTransport)!
  private var dest: any VoiceTransport { inner ?? fallback }
  func sendMessage(
    machineId: String, message: String, attachments: JsonArray?, follow: Bool
  ) async throws {
    try await dest.sendMessage(
      machineId: machineId, message: message, attachments: attachments, follow: follow)
  }
  func abandonTurn() { dest.abandonTurn() }
  func post(path: String, body: JsonObject) async throws -> JsonObject {
    try await dest.post(path: path, body: body)
  }
}

private final class DeliverBox: @unchecked Sendable {
  var deliver: @Sendable (AgentEvent) async -> Void = { _ in }
}

private final class SpeakBox: @unchecked Sendable {
  var push: @Sendable (String, String) -> Void = { _, _ in }
}

private final class SessionBox: @unchecked Sendable {
  var onState: @Sendable (AgentEvent) async -> Void = { _ in }
}

private final class HarnessRef: @unchecked Sendable {
  weak var harness: ConversationHarness?
}

public final class ConversationHarness: @unchecked Sendable {
  private let brain: any Brain
  public let clock: HarnessClock
  /// How long the model's speech occupies the channel. The window in which a nudge gets deferred.
  public var speakingMs: Int64

  private let logBox = OSAllocatedUnfairLock(initialState: [String]())
  public var log: [String] { logBox.withLock { $0 } }

  public let gate: LiveTurnGate
  public let activityLog: ActivityLog
  private let heldNudges = HeldNudgeQueue()
  public let agent: ScriptedAgent
  public let concierge: Concierge
  private let bridge: HttpAgentBridge
  private let transportBox: TransportBox
  private let me = HarnessRef()

  public private(set) var transcript: [Line] = []
  public private(set) var sessionStates: [AgentEvent] = []
  public private(set) var muted = false

  private var lastUserSpeechAt: Int64 = 0
  /// Mirrors CallService's field of the same name — the quiet clock's stop point. See `userQuietMs`.
  private var workStartedAt: Int64 = 0
  private var lastStepFailureNudgeAt: Int64 = 0

  public init(brain: any Brain, clock: HarnessClock = HarnessClock(), speakingMs: Int64 = 800) {
    self.brain = brain
    self.clock = clock
    self.speakingMs = speakingMs

    let logFn: @Sendable (String) -> Void = { [logBox] line in
      logBox.withLock { $0.append(line) }
    }
    self.gate = LiveTurnGate { clock.now }
    self.activityLog = ActivityLog { clock.now }

    let deliverBox = DeliverBox()
    let agent = ScriptedAgent(
      clock: clock,
      deliver: { await deliverBox.deliver($0) },
      log: logFn)
    self.agent = agent

    let box = TransportBox()
    box.fallback = agent
    self.transportBox = box

    let bridge = HttpAgentBridge(
      machineId: "M-harness",
      transport: box,
      log: logFn,
      nowMs: { clock.now })
    self.bridge = bridge

    let speakBox = SpeakBox()
    let voice = LiveVoiceChannel { kind, text in
      speakBox.push(kind, text)
    }

    let sessionBox = SessionBox()
    self.concierge = Concierge(
      agent: bridge,
      voice: voice,
      engine: ClientBrain(),
      timer: clock,
      onSessionState: { await sessionBox.onState($0) },
      now: { clock.now },
      log: logFn)

    sessionBox.onState = { [weak self] s in
      self?.sessionStates.append(s)
      self?.activityLog.record(agentEventToJson(s))
    }

    deliverBox.deliver = { [me] event in
      await me.harness?.onAgentEvent(event)
    }
    speakBox.push = { [me] kind, text in
      guard let harness = me.harness else { return }
      harness.clock.scheduleSuspending(delayMs: 0) {
        await harness.runGate(harness.gate.injectNudge(kind, text))
      }
    }
    me.harness = self
  }

  /// Send this run's traffic to `replacement` instead of the scripted agent. Call before `start`.
  public func useTransport(_ replacement: any VoiceTransport) {
    transportBox.inner = replacement
  }

  /// Hand an agent event to the client, as the turn-stream reader does.
  public func deliverAgentEvent(_ event: AgentEvent) async { await onAgentEvent(event) }

  /// Leave a photo on the bridge for whatever writes next — what a real capture does when it lands.
  public func stashAttachment(_ attachment: TaskAttachment) {
    bridge.addPendingAttachment(attachment)
  }

  public func state() async -> ConciergeState { await concierge.getState() }

  /// What `getSaiStatus` would answer right now — the same renderer the device uses.
  public func status() -> String { activityLog.statusText() }

  private func logLine(_ line: String) {
    logBox.withLock { $0.append(line) }
  }

  /// Note that work the user is waiting on has begun, keeping the FIRST such moment since they spoke.
  private func markWorkStarted() {
    if workStartedAt < lastUserSpeechAt || workStartedAt == 0 { workStartedAt = clock.now }
  }

  // ── Driving the conversation ───────────────────────────────────────────────

  /// Bring the Live session up.
  public func start() async {
    await runGate(gate.onConnect())
    await runGate(gate.onSetupComplete())
  }

  /// The user says something.
  public func user(_ utterance: String) async {
    lastUserSpeechAt = clock.now
    let s = await concierge.getState()
    if !s.inFlight.isEmpty || !s.queue.isEmpty { markWorkStarted() }
    transcript.append(Line(speaker: "you", text: utterance))
    _ = gate.onUserTranscript(utterance)
    await modelTurn(utterance)
  }

  /// The user talks over the model.
  ///
  /// Two things happen on a real barge-in and both matter: the server VAD raises `interrupted`, which
  /// ends the model's turn and opens the discard window, and then the new utterance is a fresh turn.
  public func bargeIn(_ utterance: String) async {
    await runGate(gate.onInterrupted())
    logLine("— barge-in —")
    await user(utterance)
  }

  /// Mute / unmute, which on the device is the temple button or the on-screen control.
  public func setMuted(_ value: Bool) async {
    muted = value
    if value {
      await runGate(gate.injectSessionState("muted", "[system] you are muted", sticky: true))
    } else {
      await runGate(gate.injectSessionState("unmuted", "[system] you are unmuted", sticky: false))
      for held in heldNudges.drain() {
        await runGate(gate.injectNudge(held.kind, held.nudge))
      }
    }
  }

  /// A token-expiry reconnect mid-call.
  public func reconnect() async {
    await runGate(gate.onConnect())
    await runGate(gate.onSetupComplete())
  }

  /// Let time pass, delivering whatever the agent had scheduled.
  public func advance(_ ms: Int64) async throws { try await clock.advance(ms) }

  /// Run until the agent has nothing left to say.
  public func settle() async throws { try await clock.drain() }

  // ── The model's half ───────────────────────────────────────────────────────

  /// Give the model a turn, and put whatever it decided through the real boundaries.
  ///
  /// The order matters and mirrors the device: the model's speech opens the turn (so a nudge arriving
  /// now is deferred by the gate), its tool calls go through the same routing GeminiLiveClient
  /// applies, and the turn closes `speakingMs` later — which is when anything held gets released.
  private func modelTurn(_ input: String) async {
    let turn = await brain.turn(input: input, state: await concierge.getState())
    if let speech = turn.speech, !speech.isEmpty {
      _ = gate.onSaiTranscript(speech)
      transcript.append(Line(speaker: "sai", text: speech))
    }
    await routeCalls(turn.calls)
    // The turn ends after the model has finished speaking, not instantly: the gap is the window in
    // which a completion landing mid-sentence is held, which is the race worth testing.
    //
    // WHICH clock matters, and getting it wrong is silent. A live transport delivers on real time and
    // nothing advances the virtual one, so scheduling the turn end there meant it never fired:
    // `modelSpeaking` stayed true for the rest of the call and every nudge was deferred behind a turn
    // that would never end. The model looked mute when it had simply never been told anything — the
    // same shape as the bug LiveTurnGate was extracted for, reproduced in the harness, and it spoiled
    // a live demo before the cause was spotted.
    let speakFor = (turn.speech?.isEmpty == false) ? speakingMs : Int64(0)
    if transportBox.inner != nil {
      try? await Task.sleep(nanoseconds: UInt64(max(speakFor, 0)) * 1_000_000)
      await runGate(gate.onGenerationOrTurnEnd(generationEnded: false, turnEnded: true))
    } else {
      clock.scheduleSuspending(delayMs: speakFor) { [me] in
        guard let harness = me.harness else { return }
        await harness.runGate(harness.gate.onGenerationOrTurnEnd(generationEnded: false, turnEnded: true))
      }
    }
  }

  /// The tool-call half of GeminiLiveClient.handleToolCall, minus the socket.
  private func routeCalls(_ calls: JsonArray) async {
    // Mirrors the real client: the gate is told a turn is in flight BEFORE any effect reaches the
    // FSM, because the FSM answers some of them by speaking and a spoken line is a client turn.
    gate.onToolCall()
    var effects: [JsonObject] = []
    let hasCapture = (0..<calls.count).contains {
      calls.optObject($0)?.optString("name") == "captureImage"
    }
    for i in 0..<calls.count {
      guard let c = calls.optObject(i) else { continue }
      let name = c.optString("name")
      let effect = fcToEffect(c)
      switch name {
      case "getSaiStatus":
        logLine("→ tool: getSaiStatus → \(status())")
      case "getLocalTime":
        logLine("→ tool: getLocalTime → \(describePhoneClock(nowMs: clock.now, timeZone: .current))")
      case "forwardToAgent", "enqueue", "relayToAgent":
        let wantsPhoto = c.optObject("args")?.optBool("attachLatestImage", false) == true
        switch gate.routeTaskCall(
          name: name, effect: effect, wantsPhoto: wantsPhoto, hasCapture: hasCapture
        ) {
        case .heldForPhoto(_, let log):
          logLine(log)
        case .emit(let log):
          effects.append(effect)
          logLine(log)
        }
      default:
        effects.append(effect)
        logLine("→ effect: \(name)")
      }
    }
    if !effects.isEmpty {
      markWorkStarted()
      _ = await concierge.applyClientEffects(JsonArray(effects.map(\.raw)))
    }
  }

  /// `{ kind: name, ...args }` — the concierge effect shape, as GeminiLiveClient builds it.
  private func fcToEffect(_ fc: JsonObject) -> JsonObject {
    var raw: [String: Any] = ["kind": fc.optString("name")]
    if let args = fc.optObject("args") {
      for (k, v) in args.raw { raw[k] = v }
    }
    return jsonWire(raw)
  }

  // ── The client's own policy on agent events (mirrors CallService.onAgentEvent)

  private func onAgentEvent(_ e: AgentEvent) async {
    // Mirrors the gate in VoiceSession's reader, and covers all three sinks for the same reason it
    // does: the tail of an aborted turn is not spoken, not written to the log `getSaiStatus` answers
    // from, and not given to the FSM. Two of those three are nothing to do with the FSM, which is why
    // suppressing its reaction alone left the result still being read out.
    if await concierge.disownsAgentEvents() {
      logLine("dropped \(eventKindForLog(e)) from an aborted turn")
      return
    }
    _ = await concierge.handleAgentEvent(e)

    let json = agentEventToJson(e)
    activityLog.record(json)
    if case .complete = e {
      transcript.append(Line(speaker: "agent", text: json.optString("summary").isEmpty
        ? json.optString("text") : json.optString("summary")))
    } else if case .error = e {
      transcript.append(Line(speaker: "agent", text: json.optString("summary").isEmpty
        ? json.optString("text") : json.optString("summary")))
    }

    let action = AgentEventRouter.route(
      event: json,
      muted: muted,
      userQuietMs: userQuietMs(now: clock.now, lastUserSpeechAt: lastUserSpeechAt, workStartedAt: workStartedAt),
      askFirstThresholdMs: defaultAskFirstMs,
      sinceLastStepFailureMs: lastStepFailureNudgeAt == 0
        ? Int64.max : clock.now - lastStepFailureNudgeAt,
      stepFailureIntervalMs: stepFailureNudgeIntervalMs)

    switch action {
    case .ignore:
      break
    case .drop(let why):
      logLine("→ nudge: \(why)")
    case .injectStepFailure(let nudge):
      lastStepFailureNudgeAt = clock.now
      await runGate(gate.injectNudge("step-failed", nudge))
    case .inject(let kind, let nudge):
      await runGate(gate.injectNudge(kind, nudge))
    case .hold(let kind, let nudge):
      if !heldNudges.add(kind: kind, nudge: nudge) {
        logLine(
          "→ nudge: \(kind) — discarded while muted (stale by the time it could be heard)")
      }
    }
  }

  // ── The gate's action interpreter ──────────────────────────────────────────

  /// Perform what the gate decided — the same job GeminiLiveClient.run does, except that a turn sent
  /// to the model comes back here as another model turn, which is what closes the loop.
  @discardableResult
  private func runGate(_ actions: [GateAction]) async -> [GateAction] {
    for action in actions {
      switch action {
      case .log(let text):
        logLine(text)
      case .sendTurn(let text):
        await modelTurn(text)
      case .saiTranscript, .userTranscript, .turnComplete, .flushPlayback:
        break
      }
    }
    return actions
  }

  // ── Assertion helpers ──────────────────────────────────────────────────────

  /// Everything the user heard, lowercased and joined — for "was this ever said" checks.
  public func heard() -> String {
    transcript.filter { $0.speaker == "sai" }.map(\.text).joined(separator: " | ").lowercased()
  }

  public func saidSomethingLike(_ fragments: String...) -> Bool {
    let h = heard()
    return fragments.contains { h.contains($0.lowercased()) }
  }

  public func logHas(_ fragment: String) -> Bool {
    log.contains { $0.contains(fragment) }
  }
}
